### PR TITLE
add a interface GraceFullShutDown to supoort customized action in shutdown

### PR DESCRIPTION
**Pull request by kele6413430**
Introduce:
In shutdown stage of a project ,gracefull shut down is necessary.
There is a application listener ,but it is not enough,some actions should be done with dubbo.
For example, Kafka producer should flush messages to the broker after protocol destroyed.
Another,MQ consumer should not poll new messages any more after registries destroyed.

Changelog

add a interface GraceFulShutDown
2.update SpringExtensionFactory.ShutdownHookListener to support customized actions when spring bean exists which implements GraceFullShutDown
3.add a unit test GraceFulShutDownTest

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [x] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/extension/GraceFulShutDown.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+    /**
+     * it is the beginning stage of shut down
+     * some resources should be closed after  dubbo registries are destroyed but before dubbo connections are destroyed
+     * so you should do something in close method
+     * at this moment,some dubbo rpc invoke may be executing now,so any destroy method for are forbidden
+     * for example, kafka message consumer should stop the poll method at this stage
+     */
+    default void afterRegistriesDestroyed(){}
```
Comment (beiwei30) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/extension/GraceFulShutDown.java]: pls. pay attention to code style, e.g, a space is required betwee ')' and '{'

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/extension/GraceFulShutDown.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * this interface is used to enable grace shutdown
+ * any class implements this interface should be a spring bean
+ * @since 2.7.3
+ */
+public interface GraceFulShutDown {
```
Comment (beiwei30) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/extension/GraceFulShutDown.java]: This interface is introduced to work and work only with spring framework. I am wondering: should we make it work for all scenarios? A quick thought is, should we make GracefulShutDown (or rename it to 'DubboLifecycle') an dubbo extension and let DubboShutdownHook to hold it?

Comment (kele6413430) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/extension/GraceFulShutDown.java]: good idea,I will edit resolve it later!

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/extension/GraceFulShutDown.java (hunk 1 above)
Comment (kele6413430) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/extension/GraceFulShutDown.java]: Get it

**Pull request closed without merging**